torch/csrc/cuda/CUDAPluggableAllocator.cpp

```diff
@@ -0,0 +1,218 @@
+    TORCH_CHECK(
+        false,
+        "CUDAPluggableAllocator does not yet support getDeviceStats. "
+        "If you need it, please file an issue describing your use case.");
+  }
+  
+  void resetAccumulatedStats(int device) override {
+    TORCH_CHECK(
+        false,
+        "CUDAPluggableAllocator does not yet support resetAccumulatedStats. "
+        "If you need it, please file an issue describing your use case.");
+  }
+  
+  void resetPeakStats(int device) override {
+    TORCH_CHECK(
+        false,
+        "CUDAPluggableAllocator does not yet support resetPeakStats. "
+        "If you need it, please file an issue describing your use case.");
+  }
+  
+  c10::cuda::CUDACachingAllocator::SnapshotInfo snapshot() override {
+    TORCH_CHECK(
+        false,
+        "CUDAPluggableAllocator does not yet support snapshot. "
+        "If you need it, please file an issue describing your use case.");
+  }
+  
+  std::shared_ptr<void> getIpcDevPtr(std::string handle) override {
+    TORCH_CHECK(
+        false,
+        "CUDAPluggableAllocator does not yet support getIpcDevPtr. "
+        "If you need it, please file an issue describing your use case.");
+
+  }
+
+  // CUDAGraph interactions
+  void notifyCaptureBegin(
+      int device,
+      c10::cuda::CaptureId_t graph_id,
+      c10::cuda::MempoolId_t mempool_id) override {
+    if (capture_begin_fn_.has_value()) {
+      capture_begin_fn_.value()(device, graph_id, mempool_id);
+    }
+  }
+  
+  void notifyCaptureAboutToEnd(int device, c10::cuda::CaptureId_t graph_id) override {
+    if (capture_about_to_end_fn_.has_value()) {
+      capture_about_to_end_fn_.value()(device, graph_id);
+    }
+  }
+
+  void notifyCaptureEnded(int device, c10::cuda::CaptureId_t graph_id) override {
+    if (capture_ended_fn_.has_value()) {
+      capture_ended_fn_.value()(device, graph_id);
+    }
+  }
+  
+  void notifyCaptureDestroy(int device, c10::cuda::MempoolId_t mempool_id) override {
+    if (capture_destroy_fn_.has_value()) {
+      capture_destroy_fn_.value()(device, mempool_id);
+    }
+  }
+
+  virtual void* raw_alloc(size_t nbytes) {
+     return nullptr;
+  }
+  virtual void* raw_alloc_with_stream(size_t nbytes, cudaStream_t stream) {
+     return nullptr;
+  }
+
+  virtual void raw_delete(void* ptr) {
+  }
+
+  virtual void recordHistory(
+      bool enabled,
+      c10::cuda::CUDACachingAllocator::CreateContextFn context_recorder,
+      size_t alloc_trace_max_entries,
+      bool alloc_trace_record_context) {
+  }
+
+  virtual void attachOutOfMemoryObserver(c10::cuda::CUDACachingAllocator::OutOfMemoryObserver observer) {
+  }
+
+  virtual bool needsPoolSpecificPeerAccess() {
+    return false;
+  }
+
+  virtual std::string name() {
+    return "pluggable";
+  }
+
+protected:
+  std::function<void*(size_t, int, cudaStream_t)> alloc_fn_;
+  std::function<void(void*)> free_fn_;
+  // TODO Populate all these functions
```

Comment:
I think for the first iteration of this, we can just support the minimal API that will make most of PyTorch work (and whatever parts you are interested in modifying). This can throw errors are return default values for stuff like:
* CUDAGraphs, mempool stuff - just error, its unlikely a custom allocator works with them unless built specific for that.
* Statistics, snapshotting - it is ok to have them return 0s, stats for other allocators are not the same anyway

Reply:
Added support for adding these functions (probably rapids guys are interested) but left it as an undocumented feature for now.